contracts/proxy-deployer/src/contract_interactions.rs

```diff
@@ -0,0 +1,206 @@
+                deployed_contracts.push(new_contract_address.clone());
+            });
+        self.deployer_template_addresses(&caller)
+            .insert(template_address, deployed_addresses);
+        self.deployers_list().insert(caller);
+
+        new_contract_address
+    }
+
+    #[endpoint(contractUpgrade)]
+    fn contract_upgrade(
+        &self,
+        contract_address: ManagedAddress,
+        template_address: ManagedAddress,
+        args: MultiValueEncoded<ManagedBuffer>,
+    ) {
+        self.can_call_endpoint(Some(contract_address.clone()));
+        require!(
+            self.blockchain().is_smart_contract(&contract_address),
+            "Contract address is not a SC"
+        );
+        require!(
+            self.blockchain().is_smart_contract(&template_address),
+            "Template address is not a SC"
+        );
+
+        self.send_raw().upgrade_from_source_contract(
+            &contract_address,
+            self.blockchain().get_gas_left(),
+            &BigUint::zero(),
+            &template_address,
+            CodeMetadata::DEFAULT,
+            &args.to_arg_buffer(),
+        );
+    }
+
+    #[endpoint(contractCallByAddress)]
+    fn contract_call_by_address(
+        &self,
+        contract_address: ManagedAddress,
+        function_name: ManagedBuffer,
+        args: MultiValueEncoded<ManagedBuffer>,
+    ) {
+        self.can_call_endpoint(Some(contract_address.clone()));
+        require!(
+            self.blockchain().is_smart_contract(&contract_address),
+            "Contract address is not a SC"
+        );
+
+        self.send()
+            .contract_call::<()>(contract_address, function_name)
+            .with_gas_limit(self.blockchain().get_gas_left())
+            .with_raw_arguments(args.to_arg_buffer())
+            .execute_on_dest_context()
+    }
+
+    #[only_owner]
+    #[allow_multiple_var_args]
+    #[endpoint(upgradeContractsByTemplate)]
+    fn upgrade_contracts_by_template(
+        &self,
+        gas_per_action: u64,
+        opt_template_address: OptionalValue<ManagedAddress>,
+        opt_args: OptionalValue<MultiValueEncoded<ManagedBuffer>>,
+    ) -> bool {
+        let mut ongoing_upgrade_operation =
+            self.get_ongoing_operation(opt_template_address, opt_args);
+
+        let default_gas_for_save = self.default_gas_for_save_operation().get();
+        while self.blockchain().get_gas_left() >= gas_per_action + default_gas_for_save
+            && !ongoing_upgrade_operation.contracts_remaining.is_empty()
+        {
+            let contract_address = ongoing_upgrade_operation
+                .contracts_remaining
+                .get(0)
+                .clone_value();
+            self.send_raw().upgrade_from_source_contract(
+                &contract_address,
+                gas_per_action,
+                &BigUint::zero(),
+                &ongoing_upgrade_operation.template_address,
+                CodeMetadata::DEFAULT,
```

Comment:
Same here. You are changing the CodeMetadata.

Reply:
The same answer as above.